### PR TITLE
Specific URL query string values should be redacted

### PR DESCRIPTION
**Pull request by trask**
Related to #860

Alternative proposal to #961

## Changes

Specific URL query string values should now be redacted by default due to concerns around leaking credentials.

The list of problematic query string keys is documented in the semantic conventions and is subject to change over time as we learn about additional common usages of credentials found in query string values.

### CODE REVIEW

**Comment by austinlparker** [comment]
Should we also add a few 'obvious' ones? e.g., 'password', 'secret', 'private', etc.?

**Comment by trask** [comment]
here's the default list used by Elastic APM: https://www.elastic.co/guide/en/apm/agent/java/1.x/config-core.html#config-sanitize-field-names

**Review by lmolkova — APPROVED** [review]
I believe we need to let users/sdks/distros to expand the blocklist via configuration, but this would be a purely incremental change.

**Comment by austinlparker** [comment]
> here's the default list used by Elastic APM: https://www.elastic.co/guide/en/apm/agent/java/1.x/config-core.html#config-sanitize-field-names

I do find it ironic that there's a big "query string values are considered non-sensitive" info box in that section. 

But yeah, that's a good list of additions too.

**Review by tylerbenson — APPROVED** [review]
I think this should be easily configured in a standard way so users can add more to it.

I'm also surprised that this list is as short as it is.

File: docs/attributes-registry/url.md [hunk 1]
```diff
@@ -26,9 +26,29 @@ linkTitle: URL
...
+
+Sensitive content provided in `url.full` SHOULD be scrubbed when instrumentations can identify it.
+
+In particular, query string values for the following keys SHOULD be redacted by default and replaced by the
+value `REDACTED`:
+
+* `AWSAccessKeyId`
+* `Signature`
+* `sig`
+* `X-Goog-Signature`
```
Comment (tylerbenson) [inline docs/attributes-registry/url.md:45]: Should we give some added context as to why these were added?  What was the decision process and what criteria should be used in the future to evaluate if something new should be added?

**Comment by gregkalapos** [comment]
I'd definitely prefer this over https://github.com/open-telemetry/semantic-conventions/pull/961, since this is less intrusive for users who already capture query strings and unlike https://github.com/open-telemetry/semantic-conventions/pull/961 this really handles values that are problematic from security point of view. With https://github.com/open-telemetry/semantic-conventions/pull/961 the only chance for people to capture query strings is to turn it back on, but that again would capture problematic values as well. So I definitely prefer an approach that tries to handle problematic values over an all or nothing config.

Since configuration came up: I think this list could be generalized and applied to multiple other things (e.g. Db query parameters, Elastic also applies this to x-form encoded request bodies (not yet part of SemConv, but could be), and there are probably many other attribtues). So the config could be a list of keys that are applied for sanitization to every attribute that can potentially contain sensitive information, not only for query strings. That could be also introduced incrementally; for now I think what the PR proposes is ok.

**Comment by reyang** [comment]
> I'm also surprised that this list is as short as it is.

If we look at what the user provided in the [original report](https://github.com/open-telemetry/opentelemetry-dotnet/security/advisories/GHSA-vh2m-22xx-q94f), here are some keys:

_It is important to ensure that sensitive query parameters are scrubbed by default. While this list is not exhaustive, the following parameters should be scrubbed as a starting point: "sig", "code", "key", "access_token", "client_secret", "password", "appkey", "app_key", "apikey", "apiKey", "api_key", "accesskey", "access_key", and "token"._


image

File: docs/attributes-registry/url.md (hunk 1 above)
Comment (reyang) [reply docs/attributes-registry/url.md:45]: +1 on put some clarification around the criteria (doesn't have to be perfect in this PR though).

I have a general concern regarding the process, for example, if this list got updated everyday, are we going to play the catch-up game from all language implementations, and do we publish security advisories and deprecate previous versions?

**Review by gregkalapos — APPROVED** [review]

File: docs/attributes-registry/url.md (hunk 1 above)
Comment (trask) [reply docs/attributes-registry/url.md]: I've added links now from each of these query string keys to the public rest APIs that use them. This is the initial criteria I'm thinking of: _known_ query string keys that are used by _known_ public rest APIs.

**Comment by trask** [comment]
There are many potential candidates for additional query string keys to include.

From initial CVE report against OpenTelemetry .Net:

* `code`
* `key`
* `access_token`
* `client_secret`
* `password`
* `appkey`
* `app_key`
* `apikey`
* `apiKey`
* `api_key`
* `accesskey`
* `access_key`
* `token`

From https://www.elastic.co/guide/en/apm/agent/java/1.x/config-core.html#config-sanitize-field-names:

* `password`
* `passwd`
* `pwd`
* `secret`
* `*key`
* `*token*`
* `*session*`
* `*credit*`
* `*card*`
* `*auth*`
* `*principal*`
* `set-cookie`

From a process perspective of how to pick and choose (both now and in the future), it might be nice to say something along the lines of

> our default redaction list only includes _known_ query string keys used to pass credentials to _known_ public rest APIs

File: docs/attributes-registry/url.md [hunk 2]
```diff
@@ -26,9 +26,29 @@ linkTitle: URL
...
+
+In particular, query string values for the following keys SHOULD be redacted by default and replaced by the
+value `REDACTED`:
+
+* [`AWSAccessKeyId`](https://docs.aws.amazon.com/AmazonS3/latest/userguide/RESTAuthentication.html#RESTAuthenticationQueryStringAuth)
+* [`Signature`](https://docs.aws.amazon.com/AmazonS3/latest/userguide/RESTAuthentication.html#RESTAuthenticationQueryStringAuth)
+* [`sig`](https://learn.microsoft.com/en-us/azure/storage/common/storage-sas-overview#sas-token)
+* [`X-Goog-Signature`](https://cloud.google.com/storage/docs/access-control/signed-urls)
+
+This list is subject to grow over time, but once a key is added to the list, removing it will be considered a
```
Comment (reyang) [inline docs/attributes-registry/url.md:47]: Why would this be considered as breaking changes?
For example, items in the current list have links to the official doc, one can imagine over years some of these technologies will be abandoned and docs would be retired.

**Review by svrnm — APPROVED** [review]

**Comment by svrnm** [comment]
I prefer this solution over #961, 👍 

Question: In [my comment](https://github.com/open-telemetry/semantic-conventions/pull/961#issuecomment-2078047508) and in [@austinlparker's comment](https://github.com/open-telemetry/semantic-conventions/pull/961#issuecomment-2079628921) and in [@reyang's comment](https://github.com/open-telemetry/semantic-conventions/pull/961#issuecomment-2079761854) (in chronological order), we discussed the idea of having generalized SDK capabilities for redaction, is this still something we want? If so I am happy to work towards a proposal for that.

**Comment by pyohannes** [comment]
> we discussed the idea of having generalized SDK capabilities for redaction, is this still something we want? If so I am happy to work towards a proposal for that.

I'd definitely answer that with a yes. Users might want to have redaction capabilities they can use for data that doesn't conform to OTel semantic conventions.

There have been related previous attempts:
* https://github.com/open-telemetry/oteps/pull/187
* https://github.com/open-telemetry/oteps/pull/100

**Comment by TylerHelmuth** [comment]
> our default redaction list only includes known query string keys used to pass credentials to known public rest APIs

I agree with adding this verbiage as it helps clarify how the list was decided.

**Review by austinlparker — APPROVED** [review]

**Comment by svrnm** [comment]
As discussed during the SemConv SIG meeting I am going to provide a proposal for the "general problem", I will try to have a draft by the end of this week to share

**Comment by svrnm** [comment]
> As discussed during the SemConv SIG meeting I am going to provide a proposal for the "general problem", I will try to have a draft by the end of this week to share

Here's the proposal: https://github.com/open-telemetry/oteps/pull/255

**Review by MSNev — APPROVED** [review]

**Comment by trask** [comment]
I've added this PR to next week's semantic convention SIG agenda to discuss resurrecting it given that https://github.com/open-telemetry/oteps/pull/255 has stalled

**Comment by svrnm** [comment]
> I've added this PR to next week's semantic convention SIG agenda to discuss resurrecting it given that [open-telemetry/oteps#255](https://github.com/open-telemetry/oteps/pull/255) has stalled

Thanks, @trask! commenting on the OTEP and why it stalled: I tried to rework it but there are so many moving pieces that I couldn't get a good handle of it, while also running into some bandwidth issues on my end, since other things required my attention. I still want to follow up on this, but it will take much more time than anticipated. So I am in favor with moving forward with this PR 👍

**Review by svrnm — APPROVED** [review]

**Review by jmacd — APPROVED** [review]

File: docs/attributes-registry/url.md (hunk 2 above)
Comment (trask) [reply docs/attributes-registry/url.md]: Resolving with https://github.com/open-telemetry/semantic-conventions/pull/971/commits/853c4851be738792f8cd5d409e049e2f78498d84

**Comment by CEbbinghaus** [comment]
One thing everybody is instantly keen on addressing is more and more keys or ways to include keys, But I would like to point out that it is equally important to allow developers to explicitly allow certain query parameters. Ultimately I think that Otel should give the developer implementing it the ultimate control of what is and isn't redacted as there will always be edge cases in either direction that cannot be addressed generically

**Pull request merged**